### ISSUE TITLE
Syntactic sugar for Task and IO to run lazy computations in pipe

### ISSUE THREAD

**Issue by magnusart** [gcanti/fp-ts#1206]
## 🚀 Feature request

I'd like some syntactic sugar for Task and IO to enable me to run the lazy computation in a pipe construct at the "end of the world". Maybe this exists but I couldn't find it.

### Current Behavior

```
const program = pipe(
stuff,
IO.map(...)
)

program() // Run program
```
or

```
pipe(
stuff,
IO.map(...)
)()
```
### Desired Behavior

```
pipe(
stuff,
IO.map(...),
IO.execute
)
```
```
pipe(
stuff,
T.map(...),
T.execute
)
```
### Suggested Solution

Add `execute` convenience method to Task and Either.

### Who does this impact? Who is this for?

All users. Makes intentions clearer I think.

### Describe alternatives you've considered

It's easy enough to write myself but it would be nice to have it in the library.

**Comment by andrewreineke** [comment]
Maybe something like this would work?

`const execute = <T>(f: () => T): T => f()`

Usage:
```
pipe(
stuff,
IO.map(...),
execute
)
```

and

```
pipe(
stuff,
T.map(...),
T.execute
)
```

**Comment by magnusart** [comment]
Yeah that would work. Not sure under what part it belongs in the library structure.

**Comment by cyberixae** [comment]
@magnusart I created a package called [ruins-ts](https://www.npmjs.com/package/ruins-ts) for this purpose.

```typescript
import * as ruins from 'ruins-ts';

ruins.fromIO(pipe(
  stuff,
  IO.map(...)
));
```

**Comment by asherccohen** [comment]
Newbie here, so here's a question.
Do pipes have lazy evaluation under the hood or any transducer mechanism?
If not, what's the best approach to optimize them in fp-ts?

**Comment by cyberixae** [comment]
@asherccohen pipes themselves have nothing lazy about them. There are some laziness in fp-ts. However, fp-ts does not provide lazy lists for example.

Pipes are convenience that lets you write `pipe(123, a, b, c, d)` instead of `d(c(b(a(123))))`. It has been [proposed](https://github.com/tc39/proposal-pipeline-operator) that a special syntax `123 |> a |> b |> c |> d` for pipes would be added to JavaScript.

There are some types in fp-ts that have laziness built into them. These would remain unexecuted unless you explicitly write `()` after the name to execute them, or use a [library](https://www.npmjs.com/package/ruins-ts) or framework that will execute them for you.

```typescript
import { Lazy } from 'fp-ts/lib/function'
import { IO } from 'fp-ts/lib/IO'
import { Task } from 'fp-ts/lib/Task'

const pure: Lazy<number> = () => 123 + 456
const sync: IO<number> = () => Math.random() * 123
const async: Task<number> = () => Promise.resolve(123)
```

Some languages like Haskell implement lazy (and potentially infinite) lists, while Python for example uses generators with [itertools](https://docs.python.org/3/library/itertools.html) utility package. The fp-ts modules are [lacking](https://github.com/gcanti/fp-ts/issues/1180) that sort of lazy evaluation. If you wish to do that you need to find another library. TypeScript does have types for [Generators](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html). However, they are not as functional as an fp-ts user might hope for.

**Comment by asherccohen** [comment]
Thanks, that explains the concepts better.
What I'm looking for is compatibility to work with big lists (if needed) using pipes and not worry that methods like map, reduce, filter iterate over the list multiple times.

Lodash used to have a lazy evaluation mechanism like that, and of course generators and observables do solve the issue too.

I am aware of the rxjs binding for fp-ts so I might reach for that in case of extreme need, I asked this question to know if there was a native way to make performant pipes or any alternative model.

**Comment by waynevanson** [comment]
@asherccohen If you want lazy evaluation, you should consider the `Iterable<A>` type, like `ixjs`.

I'm about to consider writing an fp-ts version for iterables.
